### ISSUE TITLE
Typing Booster now in Debian/Ubuntu

### ISSUE THREAD

**Issue by gunnarhj** [mike-fabian/ibus-typing-booster#122]
Typing Booster has been packaged for Debian/Ubuntu at last. In Ubuntu, version 2.9.3 of Typing Booster is available in all supported releases, while in Debian it's available in _unstable_ (and soon _testing_).

It can be installed from software management tools such as _GNOME Software_ or _Discover_, or from terminal via:

    sudo apt install ibus-typing-booster

I see at https://mike-fabian.github.io/ibus-typing-booster/ that the Ubuntu button currently points to a PPA which I set up temporarily for testing purposes. That link should be changed now, and given the URLs used for other distros, an appropriate URL might be:

https://launchpad.net/ubuntu/+source/ibus-typing-booster

### Some presentation details

Let me take this opportunity to discuss a couple of presentation details.

The whole thing gets packaged in one single binary package: `ibus-typing-booster`. That package contains these two `.desktop` files:

    /usr/share/applications/emoji-picker.desktop
    /usr/share/applications/ibus-setup-typing-booster.desktop

The first one makes Emoji Picker available via application menus and e.g. the GNOME Activities overview, while the other one is probably there to let users open the Preferences window from `gnome-control-center`. So far, so good.

When it comes to AppStream, both Typing Booster and Emoji Picker can be found via e.g. GNOME Software. Typing Booster is decently presented, but as regards Emoji Picker:

* it has the same description as Typing Booster (since there is no separate metainfo file for Emoji Picker in `/usr/share/metainfo`), and

* it refers to the very same package as Typing Booster (of course, since there is only one package), so if I click the Remove button on the Emoji Picker page, I uninstall the whole thing.

This made me wonder how it's supposed to be set up. Probably Emoji Picker ought to have its own metainfo file. Or is possibly the thought that Emoji Picker should be distributed in a separate binary package?

Finally some AppStream nitpicking. Debian checks the validity of AppStream metadata automatically, and these are the current complaints:

https://appstream.debian.org/sid/main/issues/ibus-typing-booster.html

Fixing some of those might help improve the presentation of Typing Booster.

### Thanks

Thanks for making Typing Booster available! I am honored to be able to help making it known to a broader audience.

**Comment by mike-fabian** [comment]
> Finally some AppStream nitpicking. Debian checks the validity of AppStream metadata automatically, and these are the current complaints:
> 
> https://appstream.debian.org/sid/main/issues/ibus-typing-booster.html

Do you know what tool is used to produce these warnings?

I use appstream-util to check the file like this:

```
$ appstream-util validate-relax  typing-booster.appdata.xml 
typing-booster.appdata.xml: OK
```

and it does not show warnings.

If there is a better tool for checking, I’ll use it.

**Comment by mike-fabian** [comment]
> This made me wonder how it's supposed to be set up. Probably Emoji Picker ought to have its own metainfo file. Or is possibly the thought that Emoji Picker should be distributed in a separate binary package?

On Fedora, it is in a separate binary package. On OpenSUSE, both
ibus-typing-booster and emoji-picker are in the same binary package.

Personally I preferred to have them both in the same package, but
somebody on Fedora complained about not wanting emoji-picker to be
installed but wanting ibus-typing-booster.  Not because of wanting to
save space, he did just not want extra binaries populating some gnome
menus when he doesn’t need these binaries. One cannot save much space
by not installing emoji-picker because the code and data to search for
emoji is also used by ibus-typing-booster, emoji-picker is just
another graphical frontend to use the same code to search for emoji in
a different way and to present some extra information about the emoji.

But because it was requested, I split it into two packages on Fedora.

Currently, on Fedora, there is a separate emoji-picker package which
contains just these two small files, the start script and the desktop
file:

```
$ rpm -ql emoji-picker 
/usr/bin/emoji-picker
/usr/share/applications/emoji-picker.desktop
```

Maybe I should move

```
/usr/share/ibus-typing-booster/engine/emoji_picker.py
```

as well into the emoji-picker package, looks like I somehow just
forgot this.

```
/usr/share/ibus-typing-booster/engine/itb_emoji.py
```

needs to stay in the ibus-typing-booster package as it is used both by
emoji-picker *and* by ibus-typing-booster. So the emoji-picker package
needs to require the ibus-typing-booster package.

On OpenSUSE, where I maintain the package as well, nobody complained
about that yet and therefore I didn’t make that split yet.

**Comment by mike-fabian** [comment]
> This made me wonder how it's supposed to be set up. Probably Emoji Picker ought to have its own metainfo file.

Probably I should add one with different “id”, “name”, “summary”, and “description”.

The rest of the file would be the same as ibus-typing-booster, as it
shares the same code, same home page, same issue tracker ...

**Comment by mike-fabian** [comment]
I see at https://mike-fabian.github.io/ibus-typing-booster/ that the Ubuntu button currently points to a PPA which I set up temporarily for testing purposes. That link should be changed now, and given the URLs used for other distros, an appropriate URL might be:

https://launchpad.net/ubuntu/+source/ibus-typing-booster

**Comment by gunnarhj** [comment]
>>    https://appstream.debian.org/sid/main/issues/ibus-typing-booster.html
>
> Do you know what tool is used to produce these warnings?

On Ubuntu the `appstream` package itself provides an `appstreamcli` command.

```
$ appstreamcli validate --explain /usr/share/metainfo/typing-booster.appdata.xml
W: typing-booster.xml:69: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:1791: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:205: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:622: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:976: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:688: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:425: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

E: typing-booster.xml:3: cid-is-not-rdns typing-booster.xml
    The component ID is required to follow a reverse domain-name scheme for its 
    name. See the AppStream specification for details.

W: typing-booster.xml:1085: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:115: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:1435: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:258: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:155: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:51: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

I: typing-booster.xml:3: cid-contains-hyphen typing-booster.xml
    The component ID contains a hyphen/minus. Using a hyphen is strongly 
    discouraged, to improve interoperability with other tools such as D-Bus. A good option is to 
    replace any hyphens with underscores ('_').

W: typing-booster.xml:797: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

W: typing-booster.xml:915: description-has-plaintext-url ul
    The description contains a web URL in plain text. This is not allowed, please 
    use the <url/> tag instead to share links.

Validation failed: errors: 1, warnings: 15, infos: 1
```
That might be it.

**Comment by gunnarhj** [comment]
As regards package splitting, I will follow your approach, i.e. I'll wait until someone complains.

But a metainfo file for Emoji Picker would make that application more adequately presented in e.g. GNOME Software.

**Comment by gunnarhj** [comment]
>> ... an appropriate URL might be:
>>
>> https://launchpad.net/ubuntu/+source/ibus-typing-booster
>
> Thank you very much, I changed the URL.

Great! And for the case you want to add a Debian button:

https://packages.debian.org/sid/ibus-typing-booster

**Comment by mike-fabian** [comment]
Thank you, I added a Debian button.

**Comment by mike-fabian** [comment]
How about this? Would this fix the problems with the typing-booster.appdata.xml? And is the added emoji-picker.appdata.xml file OK?

https://github.com/mike-fabian/ibus-typing-booster/commits/release-candidate-2.9.4
https://github.com/mike-fabian/ibus-typing-booster/commit/3d2bc0d05aca825d5bcb13c09a89137366c38c73
https://github.com/mike-fabian/ibus-typing-booster/commit/dbad7ecbc70fc37c6809604ea7278da39fb40ee7

**Comment by gunnarhj** [comment]
> Would this fix the problems with the typing-booster.appdata.xml? And is the added emoji-picker.appdata.xml file OK?

Both files validate on my hard disk. But I'm not aware of a sensible way to really confirm it except for uploading a new version and have the files be processed by the AppStream metadata generator.

**Comment by mike-fabian** [comment]
I'll prepare a new release ...

**Comment by mike-fabian** [comment]
https://github.com/mike-fabian/ibus-typing-booster/releases/tag/2.9.4

I hope this fixes everything mentioned in this issue.

**Comment by gunnarhj** [comment]
> I hope this fixes everything mentioned in this issue.

Not quite. The [nitpicking page](https://appstream.debian.org/sid/main/issues/ibus-typing-booster.html) is smaller, but GNOME Software now shows two Emoji Picker entries, and the new one is incorrectly shown in the _Add-ons/Input Sources_ category. I submitted a PR with some proposed changes which ought to make a difference.

**Comment by mike-fabian** [comment]
Testing your pull request ...

**Comment by gunnarhj** [comment]
Do you know of a way to do that besides an upload which triggers the AppStream generator?

**Comment by mike-fabian** [comment]
No, unfortunately not. 
I just installed an update with your pull request in a vm, gnome-software shows me ibus-typing-booster 2.9.5. 
But I cannot see the release notes and issues fixed in gnome software. 

I cannot find emoji-picker in gnome-software at all. I could not find it before I added /usr/share/metainfo/emoji-picker.appdata.xml
and I still cannot find it now.

Your pull request looks reasonable to me though. So if I cannot find a way to test it, I’ll just take it anyway, it surely does not harm.

**Comment by gunnarhj** [comment]
I don't see the release notes in GNOME Software either. My belief is that that is a design choice for GNOME Software. (In Discover I see the release notes for the latest version only.)

Maybe Fedora's setup of AppStream differs from how Debian does it.

I think that one way to test the PR without a new upstream release would be to apply the PR as a patch and upload version 2.9.4-2 to Debian. Would you like me to do that?

**Comment by mike-fabian** [comment]
I could also make a 2.9.5 release, it does not cost me much effort.

The only new things since 2.9.4 so far are:
- your appstream tweaks
- some  fixes to the Spanish translation
- somebody added a Swedish translation file but did not yet translate anything, i.e. it is still empty.

**Comment by gunnarhj** [comment]
@mike-fabian: I'd like to follow up on the AppStream changes I proposed. Commit 964f03d3 proved to not be quite correct, so I submitted #134.

This time I made an [extra Debian/Ubuntu upload](https://launchpad.net/ubuntu/+source/ibus-typing-booster/2.9.5-2) first to be sure, and yes, now both Typing Booster and Emoji Picker show up in GNOME Software (one entry each) and the Emoji Picker description is the right one. And finally no AppStream related errors or warnings.

(Without #134 Emoji Picker does not show up at all in GNOME Software.)
